otx/algorithms/segmentation/adapters/mmseg/models/losses/detcon_loss.py

```diff
@@ -0,0 +1,157 @@
+
+        def make_same_obj(ind_0, ind_1):
+            same_obj = torch.eq(
+                ind_0.reshape([batch_size, num_samples, 1]), ind_1.reshape([batch_size, 1, num_samples])
+            )
+            same_obj = same_obj.unsqueeze(2).to(torch.float)
+            return same_obj
+
+        same_obj_aa = make_same_obj(pind1, tind1)
+        same_obj_ab = make_same_obj(pind1, tind2)
+        same_obj_ba = make_same_obj(pind2, tind1)
+        same_obj_bb = make_same_obj(pind2, tind2)
+
+        # L2 normalize the tensors to use for the cosine-similarity
+        pred1 = F.normalize(pred1, dim=-1)
+        pred2 = F.normalize(pred2, dim=-1)
+        target1 = F.normalize(target1, dim=-1)
+        target2 = F.normalize(target2, dim=-1)
+
+        num_gpus = torch.cuda.device_count()
+        if num_gpus > 1 and torch.distributed.is_initialized() and self.use_replicator_loss:
+            # Grab tensor across replicas and expand first dimension
+            target1_large = [torch.zeros_like(target1) for _ in range(num_gpus)]
+            target2_large = [torch.zeros_like(target2) for _ in range(num_gpus)]
+            dist.all_gather(target1_large, target1)
+            dist.all_gather(target2_large, target2)
+            target1_large = torch.cat(target1_large, dim=0)
+            target2_large = torch.cat(target2_large, dim=0)
+
+            # Fold into batch dimension
+            target1_large = target1_large.reshape(-1, num_samples, num_features)
+            target2_large = target2_large.reshape(-1, num_samples, num_features)
+
+            # Create the labels by using the current replica ID and offsetting.
+            replica_id = dist.get_rank()
+            labels_idx = torch.arange(batch_size) + replica_id * batch_size
+            enlarged_bs = target1_large.shape[0]
+            labels = F.one_hot(labels_idx, num_classes=enlarged_bs).to(pred1.device)
+        else:
+            target1_large = target1
+            target2_large = target2
+            labels = F.one_hot(torch.arange(batch_size), num_classes=batch_size).to(pred1.device)
+
+        labels = labels.unsqueeze(dim=2).unsqueeze(dim=1)
+
+        # Do our matmuls and mask out appropriately.
+        logits_aa = torch.einsum("abk,uvk->abuv", pred1, target1_large) / self.temperature
+        logits_bb = torch.einsum("abk,uvk->abuv", pred2, target2_large) / self.temperature
+        logits_ab = torch.einsum("abk,uvk->abuv", pred1, target2_large) / self.temperature
+        logits_ba = torch.einsum("abk,uvk->abuv", pred2, target1_large) / self.temperature
+
+        labels_aa = labels * same_obj_aa
+        labels_ab = labels * same_obj_ab
+        labels_ba = labels * same_obj_ba
+        labels_bb = labels * same_obj_bb
+
+        logits_aa = logits_aa - infinity_proxy * labels * same_obj_aa
+        logits_bb = logits_bb - infinity_proxy * labels * same_obj_bb
+        labels_aa = 0.0 * labels_aa
+        labels_bb = 0.0 * labels_bb
```

Comment:
could you explain why we need to multiply by 0?

Reply:
DetCon doesn't consider features from the same view.
But, despite the same view, there must be negative samples corresponding to different masks, so we can exclude positive samples by subtracting them with very large values like L117-118 and consider both `labels_aa` and `labels_bb` as negative samples by multiplying zero.